File: apps/server1/src/server-error.ts
import { HttpErrorBase } from '@curveball/http-errors'
import httpStatus from 'http-status'

type Detail = {
  context?: unknown
  cause?: unknown
}
class ServerError extends HttpErrorBase {
  public readonly newDetail?: Detail
  public readonly statusCode: number
  public readonly isOperational: boolean

  constructor(message: string, statusCode?: number, detail?: Detail) {
    super(message)
    this.newDetail = detail
    this.statusCode = statusCode ? statusCode : httpStatus.INTERNAL_SERVER_ERROR
    this.isOperational = `${statusCode}`.startsWith('4') ? true : false
    // Error.captureStackTrace(this, this.constructor)
  }
}

export { Detail, ServerError }

File: apps/server2/src/server_error.py
from typing import Optional

from fastapi import HTTPException, status
from pydantic import BaseModel
from src.api.shared.unknown_type import UnknownType


class Detail(BaseModel):
    context: Optional[UnknownType] = None
    cause: Optional[UnknownType] = None


class ServerError(HTTPException):
    def __init__(
        self,
        message: str,
        status_code: Optional[int] = None,
        detail: Optional[Detail] = None,
    ):
        super().__init__(status_code, detail)
        self.message = message
        self.context = detail.context if detail else None
        self.cause = detail.cause if detail else None
        if status_code:
            self.status_code = status_code
        else:
            self.status_code = status.HTTP_500_INTERNAL_SERVER_ERROR
        self.is_operational = True if f"{status_code}".startswith("4") else False
        # traceback.print_stack()
        # print(repr(traceback.extract_stack()))
        # print(repr(traceback.format_stack()))
